proposals/0000-borrow-operator.md

```diff
@@ -0,0 +1,189 @@
+`pass-by-borrow` convention for that parameter instead of a `pass-by-value`
+convention. First we define the `pass-by-value` convention as follows:
+
+1. `let` and non-`inout` parameters: Formally, the compiler always copies the
+   `let` or non-`inout` parameter into a new independent value and passes the
+   value as the parameter. To the callee this value appears as an independent
+   immutable value that is different from all other parameters.
+
+2. `var` and `inout` parameters: Formally the compiler after formal evaluations
+   have completed, emits an access to the lvalue and then copies the value from
+   the lvalue into a new rvalue. The rvalue is then passed as an independent
+   value to the guaranteed parameter. Similar to the `let` case, this rvalue in
+   the callee appears as an independent immutable value that is different from
+   all other parameters.
+
+We define the `pass-by-borrow` convention as follows:
+
+1. `let` and non-inout parameters: This would cause the compiler to not emit
+   defensive copies when calling guaranteed parameters. Since a let is immutable
+   and a guaranteed parameter is immutable, we would allow for a let to be
+   passed multiple times as a `pass-by-borrow` parameter.
+
+2. `var` and `inout` parameters: This would cause the compiler to formally
+   evaluate the `var`/`inout` using a read exclusivity scope and then use a
+   borrowed formal access to pass the value to the function argument. This is
+   the same mechanism used to pass vars as an inout parameter except we have
+   exclusive read access instead of exclusive write access to the given memory.
+
+Since the `var`/`inout` case takes exclusive read access to the underlying
+memory, once cannot pass such a binding multiple times to the same function:
+
+```
+var a: Type = ...
+// Error! Cannot take exclusive access to the same variable twice
+useValueTwice(borrow a, borrow a)
+```
+
+necessarily this implies that we must also error if one passes a `var`/`inout`
+once as a `by-borrow` parameter and once as a `by-value` parameter, one will
+also achieve an exclusivity violation since one will attempt to access the
+`var`/`inout` for the lvalue-to-rvalue conversion /after/ the formal evaluation
+of the borrowed argument has begun. Example:
+
+```
+var a: Type = ...
+// Error! Cannot take exclusive access to the same variable twice
+useValueTwice(borrow a, a)
+```
+
+In pseudo-code this would look as follows:
+
+```
+var a: Type = ...
+// Formal evaluation
+let firstArgAddr = access a
+
+// RValue access
+let secondArgAddr = access a // Error! Cannot take exclusive access to the same variable twice
+let rvalue = copy secondArgAddr
+end_access secondArgAddr
+
+// Formal access (nothing to do)
+
+// Call function
+useValueTwice(firstArgAddr, rvalue)
+
+// Tear down
+end_access firstArgAddr
+```
+
+## Source compatibility
+
+`borrow` will only be allowed to be applied to lvalues (similar to `take`) and
```

Comment:
"lvalue" doesn't mean much here or in most places in the document, as a lot of things can appear on the left-side of an assignment that are probably not allowed in a `borrow`.

Reply:
I need to clean up the language here based off of https://github.com/apple/swift-evolution/blob/main/proposals/0366-move-function.md. That was my intention.